components/constellation/network_listener.rs

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use hyper::header::Location;
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+use msg::constellation_msg::PipelineId;
+use net_traits::{CoreResourceMsg, FetchMetadata, FetchResponseMsg};
+use net_traits::{IpcSend, NetworkError, ResourceThreads};
+use net_traits::request::{Referrer, RequestInit};
+use net_traits::response::ResponseInit;
+use std::sync::mpsc::Sender;
+
+pub struct NetworkListener {
+    pub res_init: Option<ResponseInit>,
+    pub req_init: RequestInit,
+    pub pipeline_id: PipelineId,
+    pub resource_threads: ResourceThreads,
+    pub sender: Sender<(PipelineId, FetchResponseMsg)>,
+    pub should_send: bool,
+}
+
+impl NetworkListener {
+    pub fn initiate_fetch(&self) {
+        let (ipc_sender, ipc_receiver) = ipc::channel().expect("Failed to create IPC channel!");
+
+        let mut req_init = self.req_init.clone();
+        if req_init.url_list.is_empty() {
+            req_init.url_list.push(req_init.url.clone());
+        }
+
+        let mut listener = NetworkListener {
+            res_init: self.res_init.clone(),
+            req_init: req_init.clone(),
+            resource_threads: self.resource_threads.clone(),
+            sender: self.sender.clone(),
+            pipeline_id: self.pipeline_id.clone(),
+            should_send: false,
+        };
+
+        ROUTER.add_route(ipc_receiver.to_opaque(), box move |message| {
+            let msg = message.to();
+            match msg {
+                Ok(FetchResponseMsg::ProcessRequestReferrer(referrer, referrer_policy)) => {
+                    listener.req_init.referrer_url = match referrer.clone() {
+                        Referrer::ReferrerUrl(servo_url) => Some(servo_url),
+                        _ => None,
+                    };
+                    listener.req_init.referrer = Some(referrer);
+                    listener.req_init.referrer_policy = referrer_policy;
```

Comment:
What if instead of the separate `ProcessRequestReferrer` message, we added the `referrer_policy` and `referrer_url` values to `Metadata`? We already include the request's final referrer value inside that structure for other code that needs it.